### PR TITLE
Fixes for Geometry Access via Python

### LINKED ISSUE #24322: ComponentInfo to python
As discovered on recent V20 experiment, we wanted to do things with component info that were not exposed to python.

1. Use iterators
2. Find a component index by name
3. In some cases (see email data), the bounding box was not computed correctly when accessed via `Shape`

See script example below:
```python
def find_component_index(name, compinfo):
    for i in range(len(compinfo)): # should be done via iterator
        if not compinfo.isDetector(i):
            if name == compinfo.name(i):
                return i
            
all = Load('/Users/spu92482/Downloads/mccode[1].h5', OutputOnlySummedEventWorkspace=False)
ws = mtd['k01_events_dat_list_p_x_y_n_id_t_all']
masked = CloneWorkspace(ws)
dinfo = masked.detectorInfo()
compinfo = masked.componentInfo()
comp_index = find_component_index('nD_Mantid_1', compinfo) # should not have to do this. indexOfAny should work
children = compinfo.detectorsInSubtree(comp_index)
min_point = compinfo.position(int(children[0]))
max_point = compinfo.position(int(children[-1]))
centre_point = (max_point + min_point) / V3D(2,2,2) # should just be able to compute average with `/2` on V3D, but not supported
for item in dinfo:
    p = item.position
    if (p.X() > centre_point.X()) or (p.Y() < centrePoint.Y()):
        item.setMasked(True)
summed = SumSpectra(masked)
rebinned = Rebin(summed, Params=[0,200,80000])
```

### PR DESCRIPTION
**Pull request by OwenArnold**
The majority of this is extending functionality in line with the problems reported. Majority of issues are on instrument/geometry side.

Fixes #24322 

**Tester**
- Review code. 
- Test out new features particularly iterators.

### CODE REVIEW

**Comment by OwenArnold** [comment]
Still TODO:
- Remove `DetectorInfoIterator.cpp`
- Move iterators from `api` to `geometry`
- Update user documentation

File: Framework/Geometry/inc/MantidGeometry/Instrument/DetectorInfoItem.h [hunk 1]
```diff
@@ -53,6 +53,12 @@ template <typename T> class DetectorInfoItem {
     return m_detectorInfo->rotation(m_index);
   }
 
+  double l2() const { return m_detectorInfo->l2(m_index); }
+
+  size_t infoSize() const { return m_detectorInfo->size(); }
```
Comment (SimonHeybrock) [inline Framework/Geometry/inc/MantidGeometry/Instrument/DetectorInfoItem.h]: This is a very unusual method to have on an iterator, reflected in the unclear name. Is it really needed?

File: Framework/Geometry/inc/MantidGeometry/Instrument/InfoIteratorBase.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  InfoItem<T> &dereference() const { return m_item; }
+
+  uint64_t distance_to(const InfoIteratorBase<T, InfoItem> &other) const {
+    return static_cast<uint64_t>(other.getIndex()) -
+           static_cast<uint64_t>(getIndex());
+  }
+
+  mutable InfoItem<T> m_item;
```
Comment (SimonHeybrock) [inline Framework/Geometry/inc/MantidGeometry/Instrument/InfoIteratorBase.h]: I feel that we discussed this before, but why does this need to be mutable? Can't the methods on `InfoItem` be const, even if they modify the underlying object (through a pointer).

File: Framework/PythonInterface/mantid/geometry/src/Exports/ComponentInfoPythonIterator.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+#endif
+          ,
+          &ComponentInfoPythonIterator::next,
+          return_value_policy<copy_const_reference>());
+  /*
+   Return value policy for next is to copy the const reference. Copy by value is
+   essential for python 2.0 compatibility because items (DetectorInfoItem) will
+   outlive their iterators if declared as part of for loops. There is no good
+   way to deal with this other than to force a copy so that internals of the
+   item are not also corrupted. Future developers may wish to choose a separte
```
Comment (SimonHeybrock) [inline Framework/PythonInterface/mantid/geometry/src/Exports/ComponentInfoPythonIterator.cpp]: What was the reason for not returning the items by value from the iterator?

File: docs/source/concepts/ComponentInfo.rst [hunk 4]
```diff
@@ -9,7 +9,28 @@ ComponentInfo
...
+
+Examples of using ``ComponentInfo`` in python
+
+Print indices of detectors in "bank1" that are masked
+
+.. code-block:: python 
+
+  from mantid.simpleapi import CreateSampleWorkspace
+  from mantid.geometry import ComponentInfo, ComponentInfoPythonIterator
+  from mantid.geometry import DetectorInfo, DetectorInfoPythonIterator
```
Comment (SimonHeybrock) [inline docs/source/concepts/ComponentInfo.rst]: Are these imports really required, even if we are not instatiating tthose objects by hand?

File: docs/source/concepts/DetectorInfo.rst [hunk 5]
```diff
@@ -9,6 +9,39 @@ DetectorInfo
...
+----------------
+
+Examples of using ``DetectorInfo`` in python
+
+Mask detectors at some distance from the source
+
+.. code-block:: python 
+
+  from mantid.simpleapi import CreateSampleWorkspace
+  from mantid.geometry import DetectorInfo, DetectorInfoPythonIterator
```
Comment (SimonHeybrock) [inline docs/source/concepts/DetectorInfo.rst]: See above.

File: docs/source/concepts/DetectorInfo.rst [hunk 6]
```diff
@@ -9,6 +9,39 @@ DetectorInfo
...
+  for item in det_info:
+      if not item.isMonitor and item.l2 > 2.0:
+          item.setMasked(True)
+
+Print detectors with scattering angle
+
+.. code-block:: python 
+
+  from mantid.simpleapi import CreateSampleWorkspace
+  from mantid.geometry import DetectorInfo, DetectorInfoPythonIterator
```
Comment (SimonHeybrock) [inline docs/source/concepts/DetectorInfo.rst]: See above.

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
Good number of tests added. 👍 

A couple of questions/comments:

File: Framework/Geometry/inc/MantidGeometry/Instrument/DetectorInfoItem.h (hunk 1 above)
Comment (OwenArnold) [reply Framework/Geometry/inc/MantidGeometry/Instrument/DetectorInfoItem.h]: Agreed. I can probably get around this.

File: Framework/Geometry/inc/MantidGeometry/Instrument/InfoIteratorBase.h (hunk 2 above)
Comment (OwenArnold) [reply Framework/Geometry/inc/MantidGeometry/Instrument/InfoIteratorBase.h]: We did discuss this before.

The restriction is imposed by `iterator_facade.hpp` specifically:
```cpp
      template <class Facade>
      static typename Facade::reference dereference(Facade const& f)
      {
          return f.dereference();
      }
```

Comment (SimonHeybrock) [reply Framework/Geometry/inc/MantidGeometry/Instrument/InfoIteratorBase.h]: Why does that require a `mutable` item?

File: Framework/PythonInterface/mantid/geometry/src/Exports/ComponentInfoPythonIterator.cpp (hunk 3 above)
Comment (OwenArnold) [reply Framework/PythonInterface/mantid/geometry/src/Exports/ComponentInfoPythonIterator.cpp]: I think that again comes in with the c++ iterator with `iterator_facade` see declaration mention above.

File: docs/source/concepts/ComponentInfo.rst (hunk 4 above)
Comment (OwenArnold) [reply docs/source/concepts/ComponentInfo.rst]: Yes, can remove those. Will do.

File: Framework/PythonInterface/mantid/geometry/src/Exports/ComponentInfoPythonIterator.cpp (hunk 3 above)
Comment (SimonHeybrock) [reply Framework/PythonInterface/mantid/geometry/src/Exports/ComponentInfoPythonIterator.cpp]: It defaults to returning `value_type &`, but I believe you can change that with an extra template argument.

File: Framework/Geometry/inc/MantidGeometry/Instrument/InfoIteratorBase.h [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+namespace Mantid {
+namespace Geometry {
+
+/** InfoIterator
+
+Base to allow users of the Info objects (DetectorInfo etc) access to data
+via a random access iterator.
+*/
+template <typename T, template <typename> class InfoItem>
+class InfoIteratorBase
```
Comment (SimonHeybrock) [inline Framework/Geometry/inc/MantidGeometry/Instrument/InfoIteratorBase.h]: I have been wondering whether we could do this much simpler: Recently I discovered `boost::transform_iterator` (see https://github.com/mantidproject/dataset/blob/master/src/dataset.h#L84). If we combine this with `boost::counting_iterator`, can we eliminate the need for a custom iterator?

```cpp
auto DetectorInfo::begin() {
  return boost::make_transform_iterator(
      boost::counting_iterator<size_t>(0),
      makeDetectorInfoItem));
}
```

Comment (SimonHeybrock) [reply Framework/Geometry/inc/MantidGeometry/Instrument/InfoIteratorBase.h]: I am not sure how `makeDetectorInfoItem` would get the instance though?

Comment (SimonHeybrock) [reply Framework/Geometry/inc/MantidGeometry/Instrument/InfoIteratorBase.h]: Can we create the lambda `makeDetectorInfoItem` using a member function? Then we can capture `this` there.

**Review by SimonHeybrock — DISMISSED** [review]

**Review by SimonHeybrock — APPROVED** [review]

File: Framework/PythonInterface/mantid/geometry/src/Exports/DetectorInfoPythonIterator.cpp [hunk 8]
```diff
@@ -20,22 +19,10 @@ void export_DetectorInfoPythonIterator() {
   // Export to Python
   class_<DetectorInfoPythonIterator>("DetectorInfoPythonIterator", no_init)
       .def("__iter__", objects::identity_function())
-      .def(
-#if PY_VERSION_HEX >= 0x03000000
-          "__next__"
+#ifdef IS_PY3K
```
Comment (soininen) [inline Framework/PythonInterface/mantid/geometry/src/Exports/DetectorInfoPythonIterator.cpp]: Would it make sense to test against Python 2.x here, leaving the `#else` clause to deal with all later Python versions? A bit of discussion on the topic can be found [here](http://astrofrog.github.io/blog/2016/01/12/stop-writing-python-4-incompatible-code/).

File: docs/source/concepts/ComponentInfo.rst [hunk 9]
```diff
@@ -9,7 +9,26 @@ ComponentInfo
...
+:ref:`Instrument Access Layers <InstrumentAccessLayers>` provides details on how `DetectorInfo` interacts with other geometry access layers.
+
+Python Interface
+----------------
+
+Examples of using ``ComponentInfo`` in python
+
+Print indices of detectors in "bank1" that are masked
+
+.. code-block:: python
```
Comment (soininen) [inline docs/source/concepts/ComponentInfo.rst]: Could this be a `.. testcode::` and a `.. testoutput::` block instead, just like in algorithm usage examples?

File: docs/source/concepts/DetectorInfo.rst [hunk 10]
```diff
@@ -9,6 +9,37 @@ DetectorInfo
...
+:ref:`Instrument Access Layers <InstrumentAccessLayers>` provides details on how `DetectorInfo` interacts with other geometry access layers.
+
+Python Interface
+----------------
+
+Examples of using ``DetectorInfo`` in python
+
+Mask detectors at some distance from the source
+
+.. code-block:: python
```
Comment (soininen) [inline docs/source/concepts/DetectorInfo.rst]: Perhaps `.. testcode::` could be considered here as well.

**Review by soininen — CHANGES_REQUESTED** [review]

**Comment by OwenArnold** [comment]
@soininen I need this in quite short order for a user. I think that your suggestions can be easily appended as part of a follow up. I've opened and assigned myself an issue to do just that.

**Review by soininen — APPROVED** [review]

**Pull request merged**